### DISCUSSION TITLE
Re-utilizing the transformation matrices from the non-linear registration

### DISCUSSION THREAD

**Discussion by jmlipman** [ANTsX/ANTs#1465] [Q&A]
Hi,

i wanted to use ANTs to perform non-linear registration and compute the transformation matrices that I will later reutilize in Python.

Specifically, I have image **I** that I want to register to template **T**, and, as a result, I get the deformed image **D**, and the files **Warp** and **InverseWarp** (note that this non-linear registration is applied after rigid+affine). This is how I do it:

`antsRegistration --dimensionality 3 
 --output [matrix_3d_nonlinear_,registered_nonlinear_3D.nii.gz] 
 --float 1 
 --interpolation Linear 
 --transform SyN[0.1,3,0] 
 --metric CC[$template3d,$brain3d,1,4] 
 --convergence [100x70x50x20,1e-6,10] 
 --shrink-factors 8x4x2x1 
 --smoothing-sigmas 3x2x1x0vox 
`

and it works like a charm. My question is, how can I re-utilize the Warp file (e.g., in Python) to transform **I** into **D**? My images are 128x128x128, and the resulting Warp file is 128x128x128x1x3. I initially thought I could convert the Warp file into voxel-wise transformation matrices, but this didn't work, probably because the content of the Warp file is something else. I also couldn't find in the documentation what are exactly the values of the Warp file (obtained via SyN). Any idea?

**Comment by ntustison** [comment]
What about using [ANTsPy](https://github.com/ANTsX/ANTsPy)?

**Reply by jmlipman** [reply]
I haven't tried it yet but I was hoping to understand how to warp the images myself so that I can implement the transformation in different programming languages. I am also planning to do some on-the-fly modifications on the warping like selecting which points of the matrix to warp and moving the warped points to a different location in space.

Maybe if I use other non-linear transformations than "SyN" it's easier to understand how the warping is done?

**Comment by ntustison** [comment]
You can look at ``antsApplyTransforms`` which utilizes the transform framework of the Insight Toolkit. Given your objectives, I would recommend asking follow-up questions over at the ITK Discourse forum.  However, if it's only to apply transforms to points, just know that that functionality is also in ANTs, ANTsPy, and ANTsR.

**Reply by jmlipman** [reply]
Thanks for redirecting me to ITK; I didn't know that ANTs saves the transformation matrices in ITK format. Thanks to that piece of information and mostly utilizing Torchio code, I wrote a simple function to apply the non-linear transformations myself. This can be useful for other people who also want to modify on-the-fly the transformations or have problems installing ANTsPy (like I do), so I leave the code here.

```python
import SimpleITK as sitk
import nibabel as nib
from torchio.data.io import sitk_to_nib, nib_to_sitk
import numpy as np

def applyWarp(nib_image: nib.Nifti1Image, warp_filepath: str):
    # warp_filepath indicates the location of the nifti file
    # created by antsRegistration via SyN "...Warp.nii.gz"

    warp = sitk.ReadImage(warp_filepath, sitk.sitkVectorFloat64)
    dis = sitk.DisplacementFieldTransform(warp)
    im_data = nib_image.get_fdata()[np.newaxis, :] # 4D

    reference = nib_to_sitk(im_data, nib_image.affine, force_3d=True)

    # Find a list of interpolators here: http://insightsoftwareconsortium.github.io/SimpleITK-Notebooks/Python_html/20_Expand_With_Interpolators.html
    interpolator = sitk.sitkLinear # Find all interpolators here:
    resampler = sitk.ResampleImageFilter()
    resampler.SetTransform(dis)
    resampler.SetReferenceImage(reference)
    resampler.SetInterpolator(interpolator)
    resampler.SetDefaultPixelValue(im_data.min())
    resampler.SetOutputPixelType(sitk.sitkFloat32)
    resampled = resampler.Execute(reference)
    res, _ = sitk_to_nib(resampled, keepdim=True)

    return res


warp_filepath = "matrix_3d_nonlinear_0Warp.nii.gz"
nibim = nib.load("registered_rigidAffine_3D.nii.gz") # 3D image

res = applyWarp(nibim, warp_filepath)

nib.save(nib.Nifti1Image(res[0], affine=nibim.affine, header=nibim.header), "output.nii.gz")
```

99% of this code is from https://github.com/fepegar/torchio/
